### PR TITLE
CG snap cephfs 7.1 feature automation - io tool and functional test w…

### PR DESCRIPTION
**Pull request by sumabai**
…orkflow

# Description
JIRAs: 
IO tool - https://issues.redhat.com/browse/RHCEPHQE-12967
Funcitonal test automation - https://issues.redhat.com/browse/RHCEPHQE-12970
Polarion TC, Test Step2- https://polarion.engineering.redhat.com/polarion/redirect/project/CEPH/workitem?id=CEPH-83581467

Latest logs(coverage for 5-10 subvolumes in quiesce set):
Test 'cg_snap_func_workflow_2' - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-BM0KVB
(This also has coverage mixed subvolumes quiesce set)

Logs: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-LZ1NTY

Note:
cg_snap_utils.py is the cg lib module used for functional test, which is due for completion.
Its is shared here for review on existing coverage - quiesce,release,query,all,include,exclude,mount and cleanup.
cephfs_cg_io.py coding and validation completed, but documentation is pending for methods, will add them by tomorrow.
cg_snap_test.py : coding and validation complete for Test Step2 in polarion TC CEPH-83581467.

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: tests/cephfs/snapshot_clone/cephfs_cg_io.py [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+        self.fs_util = FsUtils(ceph_cluster)
+
+    def start_cg_io(
+        self, clients, qs_members, client_mnt_dict, cg_test_io_status, io_run_time
+    ):
+        """
+        Start IO quiesce-set members and validate IO progress based on quiesce status
+        Args:
+            client: ceph client to mount subvolumes and run IO
+            qs_id : A list of quiesce members, each member in format subvol1 or subvol1/group1 if
```
Comment (neha-gangadhar) [inline tests/cephfs/snapshot_clone/cephfs_cg_io.py:45]: variable used in function definition and documentation is different. please add documentation for all the args of the function. The format for subvolumes of non-default group must be "group/subvolume" where group comes first and then subvolume.

File: tests/cephfs/snapshot_clone/cg_snap_test.py [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+            fs_util_v1.create_subvolume(client1, **subvolume)
+        for i in range(1, 3):
+            subvolume = {
+                "vol_name": default_fs,
+                "subvol_name": f"sv_non_def_{i}",
+                "group_name": "subvolgroup_cg",
+                "size": "6442450944",
+            }
+            fs_util_v1.create_subvolume(client1, **subvolume)
+        qs_sets = [
```
Comment (neha-gangadhar) [inline tests/cephfs/snapshot_clone/cg_snap_test.py:130]: lets have a minimum of 5 subvolumes in each quiesce set as per PM recommendations.

File: tests/cephfs/snapshot_clone/cg_snap_test.py [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+                    log.info("Verified quiesce state is QUIESCING without await option")
+                else:
+                    raise Exception(
+                        f"quiesce state of set_id {qs_id_val} is not QUIESCING"
+                    )
+                log.info("Wait for quiesced status")
+                end_time = datetime.datetime.now() + datetime.timedelta(seconds=300)
+                qs_state_verified = 0
+                while (datetime.datetime.now() < end_time) and (qs_state_verified == 0):
+                    qs_query = cg_snap_util.get_qs_query(client, qs_id=qs_id_val)
```
Comment (neha-gangadhar) [inline tests/cephfs/snapshot_clone/cg_snap_test.py:403]: can the output of "qa_query" be logged here as debug. To triage if the quiesced state is not attained within the set timeout.

File: tests/cephfs/snapshot_clone/cephfs_cg_io.py (hunk 1 above)
Comment (sumabai) [reply tests/cephfs/snapshot_clone/cephfs_cg_io.py:45]: As mentioned in description, please ignore doc section for first review, its not up-to-date, i will push the doc content by tomorrow.

File: tests/cephfs/snapshot_clone/cg_snap_test.py (hunk 2 above)
Comment (sumabai) [reply tests/cephfs/snapshot_clone/cg_snap_test.py:130]: As the 5 subvol update came later, its being considered for coding this week, its already being implemented, will validate and push changes by tomorrow.

File: tests/cephfs/snapshot_clone/cg_snap_test.py (hunk 3 above)
Comment (sumabai) [reply tests/cephfs/snapshot_clone/cg_snap_test.py:403]: noted

File: tests/cephfs/snapshot_clone/cephfs_cg_io.py (hunk 1 above)
Comment (sumabai) [reply tests/cephfs/snapshot_clone/cephfs_cg_io.py]: doc strings added to cg_snap_utils.py, cephfs_cg_io.py

File: tests/cephfs/snapshot_clone/cg_snap_test.py [hunk 4]
```diff
@@ -0,0 +1,879 @@
...
+            "FAIL: Workflow 2 - Verify quiesce release with if-version,with exclude,include prior to release"
+        )
+        return 1
+    return 0
+
+
+####################
+# HELPER ROUTINES
+####################
+def umount_all(mnt_paths, umount_params):
```
Comment (AmarnatReddy) [inline tests/cephfs/snapshot_clone/cg_snap_test.py:871]: This function is not used anywhere suma. Does this need to be used in finally block

**Review by AmarnatReddy — COMMENTED** [review]
LGTM

**Review by AmarnatReddy — COMMENTED** [review]
LGTM

File: tests/cephfs/snapshot_clone/cg_snap_test.py
Comment (hkadam134) [inline tests/cephfs/snapshot_clone/cg_snap_test.py:1]: Can the file names be renamed to "cephfs_cg_snap_test.py" - so we maintain consistency in file names

File: tests/cephfs/snapshot_clone/cg_snap_utils.py
Comment (hkadam134) [inline tests/cephfs/snapshot_clone/cg_snap_utils.py:1]: Can the file names be renamed to "cephfs_cg_snap_utils.py" - so we maintain consistency in file names

File: tests/cephfs/snapshot_clone/cg_snap_test.py (hunk 4 above)
Comment (sumabai) [reply tests/cephfs/snapshot_clone/cg_snap_test.py]: Thanks Amar, i removed this code.

**Review by neha-gangadhar — APPROVED** [review]
Approved

**Pull request merged**